Ingestion server does not appropriately report when indexer workers cannot be reached

## Description

<!-- Concisely describe the bug. Compare your experience with what you expected to happen. -->
<!-- For example: "I clicked the 'submit' button and instead of seeing a thank you message, I saw a blank page." -->

We recently encountered an issue in production wherein the indexer workers never initialized but the ingestion server was able to initialize fine. This meant that when a data refresh was initiated, the ingestion server attempted to send jobs to the indexer workers but was unable to do so since the workers (while available via DNS) were not responding to healthchecks. Here are the logs for that time period.

```
2023-07-23T00:02:30.319Z	[2023-07-23 00:02:30,319 - root - 307][INFO] Running distributed index using indexer workers.
2023-07-23T00:02:31.562Z	[2023-07-23 00:02:31,562 - root - 89][INFO] Selected worker instances ['172.31.74.186', '172.31.71.220', '172.31.71.105', '172.31.74.208', '172.31.72.237', '172.31.74.125']
2023-07-23T00:02:32.097Z	[2023-07-23 00:02:32,096 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:02:37.099Z	[2023-07-23 00:02:37,098 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:02:42.100Z	[2023-07-23 00:02:42,100 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:02:47.102Z	[2023-07-23 00:02:47,102 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:02:52.104Z	[2023-07-23 00:02:52,104 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:02:57.106Z	[2023-07-23 00:02:57,106 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:03:02.108Z	[2023-07-23 00:03:02,108 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:03:07.109Z	[2023-07-23 00:03:07,109 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:03:12.111Z	[2023-07-23 00:03:12,111 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:03:17.113Z	[2023-07-23 00:03:17,113 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:03:22.115Z	[2023-07-23 00:03:22,114 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:03:27.117Z	[2023-07-23 00:03:27,116 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:03:32.118Z	[2023-07-23 00:03:32,118 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:03:37.120Z	[2023-07-23 00:03:37,120 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:03:42.122Z	[2023-07-23 00:03:42,122 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:03:47.124Z	[2023-07-23 00:03:47,124 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:03:52.125Z	[2023-07-23 00:03:52,125 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:03:57.128Z	[2023-07-23 00:03:57,127 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:04:02.129Z	[2023-07-23 00:04:02,129 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:04:07.131Z	[2023-07-23 00:04:07,131 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:04:12.133Z	[2023-07-23 00:04:12,133 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:04:17.135Z	[2023-07-23 00:04:17,134 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:04:22.136Z	[2023-07-23 00:04:22,136 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:04:27.138Z	[2023-07-23 00:04:27,138 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:04:32.140Z	[2023-07-23 00:04:32,140 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:04:37.142Z	[2023-07-23 00:04:37,142 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:04:42.144Z	[2023-07-23 00:04:42,144 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:04:47.145Z	[2023-07-23 00:04:47,145 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:04:52.147Z	[2023-07-23 00:04:52,147 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:04:57.149Z	[2023-07-23 00:04:57,149 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:05:02.151Z	[2023-07-23 00:05:02,151 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:05:07.153Z	[2023-07-23 00:05:07,153 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:05:12.155Z	[2023-07-23 00:05:12,155 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:05:17.156Z	[2023-07-23 00:05:17,156 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:05:22.158Z	[2023-07-23 00:05:22,158 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:05:27.161Z	[2023-07-23 00:05:27,161 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:05:32.163Z	[2023-07-23 00:05:32,162 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:05:37.164Z	[2023-07-23 00:05:37,164 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:05:42.166Z	[2023-07-23 00:05:42,166 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:05:47.168Z	[2023-07-23 00:05:47,168 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:05:52.169Z	[2023-07-23 00:05:52,169 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:05:57.172Z	[2023-07-23 00:05:57,171 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:06:02.173Z	[2023-07-23 00:06:02,173 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:06:07.175Z	[2023-07-23 00:06:07,175 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:06:12.177Z	[2023-07-23 00:06:12,177 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:06:17.179Z	[2023-07-23 00:06:17,178 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:06:22.180Z	[2023-07-23 00:06:22,180 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:06:27.183Z	[2023-07-23 00:06:27,183 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:06:32.185Z	[2023-07-23 00:06:32,184 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:06:37.186Z	[2023-07-23 00:06:37,186 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:06:42.188Z	[2023-07-23 00:06:42,188 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:06:47.190Z	[2023-07-23 00:06:47,190 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:06:52.192Z	[2023-07-23 00:06:52,191 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:06:57.194Z	[2023-07-23 00:06:57,193 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:07:02.195Z	[2023-07-23 00:07:02,195 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:07:07.197Z	[2023-07-23 00:07:07,197 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:07:12.199Z	[2023-07-23 00:07:12,198 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:07:17.200Z	[2023-07-23 00:07:17,200 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:07:22.202Z	[2023-07-23 00:07:22,202 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:07:27.204Z	[2023-07-23 00:07:27,204 - root - 107][INFO] Checking http://172.31.74.186:8002/healthcheck. . .
2023-07-23T00:07:32.206Z	[2023-07-23 00:07:32,206 - root - 117][ERROR] Timed out waiting for http://172.31.74.186:8002/healthcheck.
2023-07-23T00:07:32.281Z	[2023-07-23 00:07:32,281 - root - 268][INFO] Task 42294a1d73a44b4c87edce7e76cbccb0 completed.
2023-07-23T00:07:32.286Z	[2023-07-23 00:07:32,286 - urllib3.connectionpool - 823][WARNING] Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLError(1, '[SSL: DECRYPTION_FAILED_OR_BAD_RECORD_MAC] decryption failed or bad record mac (_ssl.c:2576)'))': /api/4504934126059520/envelope/
```

## Reproduction

<!-- Provide detailed steps to reproduce the bug. -->

I was unable to reproduce this locally by stopping the indexer worker while an initialization was running (e.g. `just api/init`) because that resulted in an actual exception regarding a failed name resolution.

--- a/ingestion_server/ingestion_server/distributed_reindex_scheduler.py
+++ b/ingestion_server/ingestion_server/distributed_reindex_scheduler.py
@@ -50,11 +50,17 @@ def _assign_work(db_conn, workers, model_name, table_name, target_index):
 
     worker_url_template = "http://{}:8002"
     # Wait for the workers to start.
+    failures = []
     for worker in workers:
         worker_url = worker_url_template.format(worker)
         succeeded = _wait_for_healthcheck(f"{worker_url}/healthcheck")
         if not succeeded:
-            return False
+            failures.append(worker)
+    if failures:
+        raise ValueError(
+            f"Some workers didn't respond to health check: {','.join(failures)}"
+        )
+
     for idx, worker in enumerate(workers):
         worker_url = worker_url_template.format(worker)
         params = {

--- a/ingestion_server/test/unit_tests/test_distributed_reindex_scheduler.py
+++ b/ingestion_server/test/unit_tests/test_distributed_reindex_scheduler.py
@@ -1,0 +1,27 @@
+from unittest import mock
+
+import pytest
+
+from ingestion_server import distributed_reindex_scheduler
+
+
+def test_assign_work():
+    ...
+
+
+def test_assign_work_workers_fail():
+    mock_db = mock.MagicMock()
+    mock_db.cursor.return_value.__enter__.return_value.fetchone.return_value = [100]
+    with mock.patch(
+        "ingestion_server.distributed_reindex_scheduler._wait_for_healthcheck"
+    ) as mock_wait_for_healthcheck:
+        mock_wait_for_healthcheck.return_value = False
+
+        with pytest.raises(ValueError, match="Some workers didn't respond"):
+            distributed_reindex_scheduler._assign_work(
+                mock_db,
+                ["worker"] * 6,
+                "fake-model",
+                "fake-table",
+                "target-index",
+            )

--- a/ingestion_server/test/unit_tests/test_distributed_reindex_scheduler.py
+++ b/ingestion_server/test/unit_tests/test_distributed_reindex_scheduler.py
@@ -1,12 +1,72 @@
+import json
 from unittest import mock
 
+import pook
 import pytest
 
 from ingestion_server import distributed_reindex_scheduler
 
 
-def test_assign_work():
-    ...
+@pytest.mark.parametrize(
+    "estimated_records, record_limit, workers, expected_ranges",
+    [
+        # One worker
+        (100, 1000, ["worker1"], [(0, 100)]),
+        # Multiple workers, even split
+        (100, 1000, ["worker1", "worker2"], [(0, 50), (50, 100)]),
+        # Multiple workers, uneven split
+        # NOTE: records get dropped in this case
+        (100, 1000, ["worker1", "worker2", "worker3"], [(0, 33), (33, 66), (66, 99)]),
+        # One worker, limited
+        (100, 55, ["worker1"], [(0, 55)]),
+        # Two workers, limited
+        (100, 50, ["worker1", "worker2"], [(0, 25), (25, 50)]),
+    ],
+)
+def test_assign_work(estimated_records, record_limit, workers, expected_ranges):
+    # Checks for the parameters
+    assert len(workers) == len(
+        expected_ranges
+    ), "Number of workers and expected ranges do not match, correct the test parameters"
+    # Set up database mock response
+    mock_db = mock.MagicMock()
+    mock_db.cursor.return_value.__enter__.return_value.fetchone.return_value = [
+        estimated_records
+    ]
+    # Enable pook & mock other internal functions
+    with pook.use(), mock.patch(
+        "ingestion_server.distributed_reindex_scheduler.get_record_limit"
+    ) as mock_get_record_limit, mock.patch(
+        "ingestion_server.distributed_reindex_scheduler._wait_for_healthcheck"
+    ) as mock_wait_for_healthcheck:
+        mock_wait_for_healthcheck.return_value = True
+        mock_get_record_limit.return_value = record_limit
+
+        post_mocks = [
+            pook.post(f"http://{worker}:8002/indexing_task") for worker in workers
+        ]
+
+        distributed_reindex_scheduler._assign_work(
+            mock_db,
+            workers,
+            "sample_model",
+            "sample_table",
+            "sample_index",
+        )
+
+        for worker, mock_post, (start_id, end_id) in zip(
+            workers, post_mocks, expected_ranges
+        ):
+            assert mock_post.calls > 0
+            data = json.loads(mock_post.matches[0].body)
+            print(f"{data=}")
+            assert data == {
+                "model_name": "sample_model",
+                "table_name": "sample_table",
+                "target_index": "sample_index",
+                "start_id": start_id,
+                "end_id": end_id,
+            }
 
 
 def test_assign_work_workers_fail():
@@ -14,7 +74,7 @@ def test_assign_work_workers_fail():
     mock_db.cursor.return_value.__enter__.return_value.fetchone.return_value = [100]
     with mock.patch(
         "ingestion_server.distributed_reindex_scheduler._wait_for_healthcheck"
-    ) as mock_wait_for_healthcheck:
+    ) as mock_wait_for_healthcheck, pook.use(False):
         mock_wait_for_healthcheck.return_value = False
 
         with pytest.raises(ValueError, match="Some workers didn't respond"):

--- a/ingestion_server/test/unit_tests/test_distributed_reindex_scheduler.py
+++ b/ingestion_server/test/unit_tests/test_distributed_reindex_scheduler.py
@@ -1,4 +1,3 @@
-import json
 from unittest import mock
 
 import pook
@@ -42,9 +41,17 @@ def test_assign_work(estimated_records, record_limit, workers, expected_ranges):
         mock_wait_for_healthcheck.return_value = True
         mock_get_record_limit.return_value = record_limit
 
-        post_mocks = [
-            pook.post(f"http://{worker}:8002/indexing_task") for worker in workers
-        ]
+        # Set up pook matches
+        for worker, (start_id, end_id) in zip(workers, expected_ranges):
+            pook.post(f"http://{worker}:8002/indexing_task").json(
+                {
+                    "model_name": "sample_model",
+                    "table_name": "sample_table",
+                    "target_index": "sample_index",
+                    "start_id": start_id,
+                    "end_id": end_id,
+                }
+            )
 
         distributed_reindex_scheduler._assign_work(
             mock_db,
@@ -54,19 +61,8 @@ def test_assign_work(estimated_records, record_limit, workers, expected_ranges):
             "sample_index",
         )
 
-        for worker, mock_post, (start_id, end_id) in zip(
-            workers, post_mocks, expected_ranges
-        ):
-            assert mock_post.calls > 0
-            data = json.loads(mock_post.matches[0].body)
-            print(f"{data=}")
-            assert data == {
-                "model_name": "sample_model",
-                "table_name": "sample_table",
-                "target_index": "sample_index",
-                "start_id": start_id,
-                "end_id": end_id,
-            }
+        # Pook will raise an exception here if any requests don't match the above
+        assert pook.isdone()
 
 
 def test_assign_work_workers_fail():

--- a/ingestion_server/test/unit_tests/test_distributed_reindex_scheduler.py
+++ b/ingestion_server/test/unit_tests/test_distributed_reindex_scheduler.py
@@ -61,7 +61,7 @@ def test_assign_work(estimated_records, record_limit, workers, expected_ranges):
             "sample_index",
         )
 
-        # Pook will raise an exception here if any requests don't match the above
+        # Raise an exception to ensure all pook requests were matched
         assert pook.isdone()
 
 

